File: python/remove_element.py
# https://leetcode.com/explore/learn/card/fun-with-arrays/526/deleting-items-from-an-array/3247/

# O(n)
class Solution:
    def removeElement(self, nums: List[int], val: int) -> int:
        head = 0
        tail = len(nums) - 1
        
        while head <= tail:
            if nums[tail] == val:
                tail -= 1
                continue
            if nums[head] == val:
                x = nums[head]
                nums[head] = nums[tail]
                nums[tail] = x
                tail -= 1
            head += 1
        return tail + 1

File: java/RemoveElement.java
// https://leetcode.com/explore/learn/card/array-and-string/205/array-two-pointer-technique/1151/

// O(n)
class Solution {
    public int removeElement(int[] nums, int val) {
        int slow = -1, fast = 0;
        
        while(fast < nums.length) {
            if(nums[fast] != val) {
                nums[++slow] = nums[fast];
            }
            fast++;
        }
        
        return ++slow;
    }
}
